src/image/TransmissionCurve.cc

```diff
@@ -0,0 +1,851 @@
+#include "lsst/afw/table/io/OutputArchive.h"
+#include "lsst/afw/table/io/CatalogVector.h"
+
+#define LSST_CHECK_GSL(type, status) \
+    if (status) throw LSST_EXCEPT(type, std::string("GSL error: ") + ::gsl_strerror(status))
+
+
+namespace lsst {
+namespace afw {
+namespace image {
+
+namespace {
+
+template <typename T>
+using CPtr = std::unique_ptr<T, void (*)(T*)>;
+
+
+double computeNaturalSamplingSize(ndarray::Array<double,1,1> const & wavelengths) {
+    double last = -1.0;
+    double minSpacing = -1.0;
+    for (auto wl : wavelengths) {
+        if (last < 0) {
+            last = wl;
+        } else {
+            double spacing = wl - last;
+            if (spacing < 0) {
+                throw LSST_EXCEPT(
+                    pex::exceptions::InvalidParameterError,
+                    "Input array of wavelengths is not monotonically increasing."
+                );
+            }
+            if (minSpacing < 0 || spacing < minSpacing) {
+                minSpacing = spacing;
+            }
+            last = wl;
+        }
+    }
+    // We round instead of (conservatively) using ceil to correctly handle the case where the given
+    // wavelength sampling *is* even, but round-off error makes minSpacing *slightly* too small
+    // for exact division.
+    return 1 + std::round((wavelengths.back() - wavelengths.front())/minSpacing);
+}
+
+
+class IdentityTransmissionCurve : public TransmissionCurve {
+public:
+
+    // IdentityTransmissionCurve is a singleton.
+    static std::shared_ptr<IdentityTransmissionCurve> get() {
+        static std::shared_ptr<IdentityTransmissionCurve> instance(new IdentityTransmissionCurve());
+        return instance;
+    }
+
+    SampleDef getNaturalSampling() const override {
+        return SampleDef();
+    }
+
+    std::pair<double,double> getThroughputAtBounds() const override {
+        return std::make_pair(1.0, 1.0);
+    }
+
+    void sampleAt(
+        geom::Point2D const &,
+        ndarray::Array<double const,1,1> const & wavelengths,
+        ndarray::Array<double,1,1> const & out
+    ) const override {
+        out.deep() = 1.0;
+    }
+
+    bool isPersistable() const override { return true; }
+
+protected:
+
+    // transforming an IdentityTransmissionCurve is a no-op
+    std::shared_ptr<TransmissionCurve> _transformImpl(
+        std::shared_ptr<TransmissionCurve> self,
+        std::shared_ptr<geom::TransformPoint2ToPoint2> transform
+    ) const override {
+        return self;
+    }
+
+    // multiplying an IdentityTransmissionCurve always yields the other operand
+    std::shared_ptr<TransmissionCurve> _multiplyImpl(
+        std::shared_ptr<TransmissionCurve> self,
+        std::shared_ptr<TransmissionCurve> other
+    ) const {
+        return other;
+    }
+
+    std::string getPersistenceName() const override { return "IdentityTransmissionCurve"; }
+
+    std::string getPythonModule() const override { return "lsst.afw.image"; }
```

Comment:
Consider using string literals, now that we're in C++14 (though the compiler would probably convert them anyway).